tracing/src/macros.rs

```diff
@@ -632,35 +630,38 @@ macro_rules! event {
         $crate::event!(target: $target, parent: $parent, $lvl, { $($arg)+ })
     );
     (target: $target:expr, $lvl:expr, { $($fields:tt)* } )=> ({
-        $crate::__tracing_log!(
+        use $crate::__macro_support::Callsite as _;
+        static CALLSITE: $crate::__macro_support::MacroCallsite = $crate::callsite2! {
+            name: concat!(
+                "event ",
+                file!(),
+                ":",
+                line!()
+            ),
+            kind: $crate::metadata::Kind::EVENT,
             target: $target,
-            $lvl,
-            $($fields)*
-        );
-        if $crate::level_enabled!($lvl) {
-            use $crate::__macro_support::Callsite as _;
-            static CALLSITE: $crate::__macro_support::MacroCallsite = $crate::callsite2! {
-                name: concat!(
-                    "event ",
-                    file!(),
-                    ":",
-                    line!()
-                ),
-                kind: $crate::metadata::Kind::EVENT,
+            level: $lvl,
+            fields: $($fields)*
+        };
+        (|value_set: $crate::field::ValueSet| {
+            $crate::__tracing_log!(
                 target: $target,
-                level: $lvl,
-                fields: $($fields)*
-            };
-            let interest = CALLSITE.interest();
-            if !interest.is_never() && CALLSITE.is_enabled(interest)  {
-                let meta = CALLSITE.metadata();
-                // event with contextual parent
-                $crate::Event::dispatch(
-                    meta,
-                    &$crate::valueset!(meta.fields(), $($fields)*)
-                );
+                $lvl,
+                value_set
+            );
+            if $crate::level_enabled!($lvl) {
```

Comment:
Hmm. It looks like the new code will now _always_ construct the `ValueSet`, even if the event is not enabled _and the `log` feature is disabled_. This seems like it could have a potential performance impact. I think we should ensure that the `ValueSet` is constructed only as needed.

I think we might be able to achieve that by changing the `if` structure to something like

```rust
if $crate::level_enabled!($lvl) && { 
       let interest = CALLSITE.interest();
        !interest.is_never() && CALLSITE.is_enabled(interest)
    }
{
   let value_set = $crate::valueset!(...);
   // tracing stuff
   $crate::__tracing_log!( ... value_set);
} else {
    $crate::__tracing_log!( ..., $crate::valueset!(...))
}
```

Also, it's probably worth running the benchmarks on the `master` branch, and then running them again on this branch, to see how much of a performance impact this change has?

Reply:
That does look possible, I'll have a go. I initially focused on expanding valueset only once because that's obviously correct, but two branches of the same if statement should also work I think (essentially as long as the compiler can prove only one will ever execute).

Reply:
Done, works fine.